### DISCUSSION TITLE
JTE Gradle Plugin issues

### DISCUSSION THREAD

**Discussion by leonard84** [casid/jte#334] [Q&A]
I've noticed that the plugin puts the precompiled classes in `<project>/jte-classes` instead of `<project>/build/jte-classes`, as is customary for any generated output. 
Is there a specific reason for that, or can this be fixed?

While looking at the plugin code I've noticed some other issues:
- `JteTaskBase` is accessing the extension to read and store configuration. This does not comply with idiomatic plugin design in Gradle. A `Task` should have its own properties, which are then wired up using `prop.convention(extension.prop)` for the default registered task via the plugin. This makes it possible to declare another task with a different configuration.
- The `JteExtension` uses `Property<Path>` instead of `DirectoryProperty` to for properties meant as directories

Fixing the latter can't easily be done in a backward-compatible way. 
Would you accept pull requests that address these issues?

**Comment by casid** [comment]
Hi @leonard84,

sorry for the late reply and thank you for the suggestions. Unfortunately, I'm not very experienced with Gradle, so can't really grasp the impact of these changes. @iRebbok has written the first version of the Gradle plugin and @edward3h has modernized it to the latest Gradle best practices (by then :-)). Maybe one of you could have a look at the proposed changes?

As for your first question, I think the target directory for the precompiled classes is set [here](https://github.com/casid/jte/blob/5e206e155bff549ad25c59ae2aaa1d9bba36b945/jte-gradle-plugin/src/main/java/gg/jte/gradle/PrecompileJteTask.java#L86) and can be configured by the user by setting the `targetDirectory` property.

**Comment by edward3h** [comment]
Hi @leonard84 ,
I did an update to the Gradle plugin a while back. Your suggestions sound fine to me. Feel free to make a PR and tag me for review.

For the directory properties it would be good to maintain backward compatibility. Maybe that is possible with wrapper methods. If not it will be a very small fix to make on an affected build file.

**Comment by edward3h** [comment]
@casid I've seen a few issues relating to the Gradle plugin. Could you add a label "gradle" to them?

**Reply by casid** [reply]
Will do!

**Comment by edward3h** [comment]
I merged a change https://github.com/casid/jte/pull/478 a while back. This maintains compatibility, while allowing you to declare a new task that is independent of the default ones. @casid would you mind doing a release?

As @leonard84 mentioned, I couldn't migrate `Property<Path>` to `DirectoryProperty` and still maintain backward compatibility. I really tried!

**Reply by casid** [reply]
@edward3h I hope I can do one this weekend.

**Reply by casid** [reply]
Version 3.2.2 is released.

This is the first release using the new central publishing, it took quite a while! Built and uploaded everything yesterday evening and eventually went to bed (was in status publishing for about an hour). But was published when I woke up today 👍
